vtkext/private/module/vtkF3DRenderer.cxx

```diff
@@ -518,26 +599,30 @@ void vtkF3DRenderer::ConfigureGridUsingCurrentActors()
         tmpUnitSquare = pow(10.0, round(log10(diag * 0.1)));
       }
 
-      double gridPos[3] = { 0, 0, 0 };
+      double center[4] = { 0, 0, 0, 1 };
+      bbox.GetCenter(center);
+
+      double* gridPos = upMatrixInv->MultiplyDoublePoint(center);
+
+      double downShift = 0;
       if (this->GridAbsolute)
       {
-        for (int i = 0; i < 3; i++)
-        {
-          gridPos[i] = this->UpVector[i] ? 0 : 0.5 * (bounds[2 * i] + bounds[2 * i + 1]);
-        }
+        double origin[3] = { 0, 0, 0 };
+        downShift += vtkPlane::DistanceToPlane(center, up, origin);
       }
       else
       {
-        for (int i = 0; i < 3; i++)
-        {
-          // a small margin is added to the size to avoid z-fighting if large translucent
-          // triangles are exactly aligned with the grid bounds
-          constexpr double margin = 1.0001;
-          double size = margin * (bounds[2 * i + 1] - bounds[2 * i]);
-          gridPos[i] = 0.5 * (bounds[2 * i] + bounds[2 * i + 1] - this->UpVector[i] * size);
-        }
+        // a small margin is added to the size to avoid z-fighting if large translucent
+        // triangles are exactly aligned with the grid bounds
+        constexpr double margin = 0.0001;
+        downShift += bbox.GetLength(1) / 2 + margin;
       }
 
+      double delta[3];
+      this->GetEnvironmentUp(delta);
+      vtkMath::MultiplyScalar(delta, downShift);
+      vtkMath::Subtract(gridPos, delta, gridPos);
```

Comment:
not sure here, but it seems like you are using the 4x4 matrix to rotate and then this code does the translation.  
You can encode rotation and translation within the same 4x4 matrix (last column for the translation)